Add a document describing firewall options

puneetse opened the pull request:
This PR adds a document describing firewall options in CL.

Closes https://github.com/clearlinux/clear-linux-documentation/issues/541

ahkok commented on source/clear-linux/guides/network/firewalls.rst:
```diff
@@ -0,0 +1,198 @@
+
+|CL| does not impose a firewall policy out of the box. All traffic is allowed
+inbound and all traffic is allowed outbound. However, `tallow`_ is installed
+by default and may dynamically create rules temporarily restricting SSH
```
..`may dynamically insert one rule temporarily`...

ahkok replied:
`restricting all access` - not just SSH.

ahkok commented on source/clear-linux/guides/network/firewalls.rst:
```diff
@@ -0,0 +1,198 @@
+   misconfigured. If this happens on a remote host, local recovery may be
+   required.
+   
+   Be sure to test your firewall configuration before committing it permanently.
```
perhaps "and ensure your system will remain accessible remotely in case this is a requirement." ?

ahkok commented on source/clear-linux/guides/network/firewalls.rst:
```diff
@@ -0,0 +1,198 @@
+   .. code:: bash
+
+      #Delete all existing rules 
+      sudo iptables -F
```
I wouldn't do this by default. Definitely note that some programs may insert rules and expect this not to be done randomly.

ahkok commented on source/clear-linux/guides/network/firewalls.rst:
```diff
@@ -0,0 +1,198 @@
+
+`ipset <http://ipset.netfilter.org/>`_ is a framework in the Linux kernel for
+storing and efficiently indexing combinations of IP addresses, networks,
+(TCP/UDP) port numbers, MAC addresses, and interface names.
```
"By itself, `ipset` does not enforce network traffic rules. It acts as a simple but efficient table that allows `netfilter` rules to match network packets against a large set of entries. This makes writing network policy rules simpler, but also a lot more efficient."

ahkok commented on source/clear-linux/guides/network/firewalls.rst:
```diff
@@ -0,0 +1,198 @@
+
+Custom ipsets do not need to be defined to use iptables but, iptables can be
+extended with IP sets for matching against a large and/or changing sets of
+hosts.
```
Note that in order for `netfilter` rules to be inserted, the `ipset` must be created before a `netfilter` rule can match against them. In other words, you need to create the `ipset` rule before you can reference it in a `netfilter` rule.

ahkok commented on source/clear-linux/guides/network/firewalls.rst:
```diff
@@ -0,0 +1,198 @@
+      sudo swupd bundle-add firewalld
+
+
+#. Disable *iptables*, *ipset*, and *tallow* services as they conflict with
```
Technically not entirely correct. Better would be to modify `tallow.service` to make it *not* create iptables rules - it has an option for that, and have firewalld insert this rule. `man tallow` should document this.

puneetse replied:
I created https://github.com/clearlinux/tallow/issues/11 to add firewalld support to tallow. 

> and have firewalld insert this rule. man tallow should document this.

I'm uncertain if the `firewalld.ipset` firewalld uses is the same underlying structure as the regular `ipset`. https://firewalld.org/documentation/man-pages/firewalld.ipset.html . So even if the firewalld rule is manually created, I don't think tallow would be updating the same data structure.

```
 $ sudo firewall-cmd --get-ipsets
```

```
$ sudo ipset list
Name: tallow
Type: hash:ip
Revision: 4
Header: family inet hashsize 1024 maxelem 65536 timeout 3600
Size in memory: 88
References: 0
Number of entries: 0
Members:

Name: tallow6
Type: hash:ip
Revision: 4
Header: family inet6 hashsize 1024 maxelem 65536 timeout 3600
Size in memory: 104
References: 0
Number of entries: 0
Members:
```

So my understanding is that in this configuration tallows list wouldn't be used anyways, hence I disabled it. Can you confirm my understanding?

ahkok replied:
if `ipset` doesn't show firewalld sets then something is really wrong. I don't know what to do in that case - ipsets can't be "temporarily missing" because iptables will not allow you to insert rules referring to it - because it does not exist.

puneetse replied:
So, in its current state doesn't it make sense to disable tallow if firewalld is chosen for use? 
tallow will be creating ipsets/rules that firewalld cannot not use. 

The net result is the same whether tallow is disabled or if is configured not to create rules, right?

ahkok replied:
Disabling tallow never makes sense, imho. We should absolutely drive a solution that doesn't require users to `if (want_feature_x) { disable(feature_y) }`.

If firewalld is stubborn about it, we should patch it to not break tallow (functionally). Especially if firewalld doesn't appear to be able to behave.

Maybe we need to pre-configure firewalld so it runs *after* tallow and somehow happily lives along with tallow.



> The net result is the same whether tallow is disabled or if is configured not to create rules, right?

Not even close? If tallow doesn't make rules, but -say- iptables-restore.service does, tallow is (essentially) fully functional. If you disable tallow, it is entirely non-functional.

puneetse replied:
> If tallow doesn't make rules, but -say- iptables-restore.service does, tallow is (essentially) fully functional.

But the suggestion here is to mask iptables-restore if firewalld is being used so this case wouldn't occur.

In any case, I'll update the document to create a tallow.conf with `nocreate` instead.

ahkok commented on source/clear-linux/guides/network/firewalls.rst:
```diff
@@ -0,0 +1,198 @@
+  network port you're expecting with :command:`lsof` or :command:`netstat`.
+
+* For outbound connections, make sure the destination host is responding to
+  the network port you're expecting with :command:`telnet`. If the connection
```
better: `ping`. Definitely not `telnet`. even `nmap` would be better, I feel.

puneetse commented:
Thanks for the feedback and conversation @ahkok. I've pushed an updated version incorporating the feedback but have an open regard tallow. Could you have a second look?

mvincerx commented:
LGTM. @ahkok do you have any other comments--esp on tallow config?

ahkok commented:
Yes, I don't see or haven't seen anything that shows that `firewalld` will create the needed tallow ipsets.

Passing the `nocreate` flag to tallow will make it omit creating iptables and ipsets. So, you need to at least show how to configure firewalld to create these ipsets and the needed iptables lines, otherwise you should instruct users to disable tallow, because it will just produce errors and do nothing useful.

ahkok commented:
Note: with `nocreate` tallow will still attempt to insert ipset entries when it needs to block a host temporarily.

puneetse commented:
I'm looking into the prerequisite issues with firewalld on CL and tallow integration/documentation. Please hold on this PR.

puneetse commented:
I've rebased this with master and removed the content suggesting to disable tallow. 

Instead, firewalld support has been manufactured for tallow here: https://github.com/clearlinux/tallow/pull/12 and other firewalld bugs resolved here: https://github.com/clearlinux/distribution/issues/1123.

puneetse commented:
@ahkok can you do another review of this document? I think we can safely publish this now with clearlinux/tallow#12 following separately.

ahkok commented on source/guides/network/firewall.rst:
```diff
@@ -0,0 +1,204 @@
+   <https://firewalld.org/documentation/>`_ to learn more about
+   firewalld concepts and configuration options.
+
+   Below is a common example to allow HTTP traffic in public zones:
```
We should default to `HTTPS` here for the example. Not insecure HTTP.

mvincerx replied:
Waiting on final edit. Will merge after your requested change is made.

puneetse replied:
Done.

mvincerx replied:
Cheers.

ahkok commented:
Yeah, I think this is nice and simple now. Let's go and push this.

mvincerx commented:
LGTM
